File: cpn_linearize_basics.py
import numpy as np
import scipy as sp
import time

import memory_profiler as mp

@mp.profile
def main():
    # %% Start file for TenSyGrid Hackathon 16.12.2025
    # Linearization for iMTI models in CPN representation (basic, functional)

    # Configuration
    check = 0  # 0 = sparse/scalable/fast | 1 = full (only for checks)
    test = False
    test_zeros = False
    debug = True
    # %% Define problem


    # iMTI model in CPN representation
    if test:
        n = 2               # number of states
        m = 1               # number of inputs
        p = 0               # number of outputs
        q = n               # number of equations
        N = 2 * n + m + p   # total number of signals

        S = np.array([
            [1, 0, 0, 0, 0, 0],
            [0, 1, 0, 0, 0, 0],
            [0, 0, 1, 0, 0, 1],
            [0, 0, 0, 1, 0, 1],
            [0, 0, 0, 0, 1, 0],
        ]) # Structure matrix
        P = np.array([
            [-1, 0, -1, 0, 1, 0],
            [0, -1, 0, -1, 0, 1],
        ]) # Parameter matrix
        r = S.shape[1] # rank (number of non-zero elements in the structure matrix)
        S= sp.sparse.csr_matrix(S)
        P = sp.sparse.csr_matrix(P)

        # Linearization point
        if not test_zeros:
            dx = np.ones((n, 1))    # state derivative 
        else:
            dx = np.zeros((n, 1))   # (CONFLICTIVE CASE) state derivative (TODO: Develop)
        x = 2*np.ones((n, 1))       # state
        u = np.ones((m, 1))         # input
        y = np.ones((p, 1))         # output

    else:
        n = 1000                    # number of states
        m = 1                       # number of inputs
        p = 0                       # number of outputs
        q = n                       # number of equations
        N = 2 * n + m + p           # total number of signals
        r = 1000                       # rank (number of non-zero elements in the structure matrix)
        S = np.random.rand(N, r)    # Structure matrix
        S = np.round(2*S-1)
        P = np.random.rand(q, r)    # Parameter matrix
        P = P*(P<0.3)
        S = sp.sparse.csr_matrix(S)
        P = sp.sparse.csr_matrix(P)

        # linearization point
        dx = np.random.rand(n, 1)   # state derivative 
        x = np.random.rand(n, 1)    # state
        u = np.random.rand(m, 1)    # input
        y = np.random.rand(p, 1)    # output

    # signal vector
    v = np.vstack([dx, x, u, y])        # signal vector

    tic = time.process_time()   # start the cputime clock
    # %% Linearization (Scalable = sparse & low rank)
    if debug: print(S.toarray())
    X = S.multiply(v) - abs(S)
    if debug: print(X.toarray())
    icrit = X == -1 # get numerical critical indices
    if icrit.nnz > 0:
        raise RuntimeError("specials to be implemented ?")

    X.data = X.data + 1.0 # add one only to the nonzero elements

    # We want, for all S != 0 and X == 0, to set X to 1 efficiently (without full matrix loops).
    # Both S and X are sparse.
    # So: find the (i,j) where S is nonzero but X is 0, and efficiently set X[i,j]=1.
    S_nonzero = S.nonzero()
    X_nz_dict = set(zip(*X.nonzero()))

    # Collect new (i,j) that are in S but *not* in X (i.e., S!=0 and X==0)
    to_add = []
    for i, j in zip(*S_nonzero):
        if (i, j) not in X_nz_dict:
            to_add.append((i, j))

    if to_add:  # if there are positions to modify
        # prepare data to add (location and value = 1.0)
        rows, cols = zip(*to_add)
        data = np.ones(len(to_add))
        shape = X.shape
        X_add = sp.sparse.coo_matrix((data, (rows, cols)), shape=shape)
        X = X + X_add
        X = X.tocsr()  # ensure CSR format for downstream code

    if debug: print(X.toarray())

    # get the indices and values of X
    rowi, coli = X.nonzero()
    val = X.data

    Y = np.zeros(r) # initialize product vector
    for c in range(r): # product over nonzero elements
        mask = coli == c # get the indices of the non-zero elements in the column c
        if np.any(mask): # if there are non-zero elements in the column c
            Y[c] = np.prod(val[mask]) # compute the product of the non-zero elements in the column c

    X.data = 1.0 / X.data # Invert only nonzero elements of X
    F = S.multiply(Y).multiply(X)

    EABC = P*(F.T) 
    E = -EABC[:, 0:n] # state matrix
    A = EABC[:, n:2*n] # system matrix
    B = EABC[:, 2*n:2*n+m] # input matrix

    # Local stability: dominant (geralized) eigenvalues
    A_dense = A.toarray()
    E_dense = E.toarray()

    #eigvals, eigvecs = sp.linalg.eig(A_dense, E_dense)  # resuelve A v = λ E v
    #print(eigvals)
    toc = time.process_time()
    print( "Time: ", toc-tic)

if __name__ == "__main__":
    mp.profile(main())

File: cpn_linearize_basics.m
%% Start file for TenSyGrid Hackaton 16.12.2025
% linearization for iMTI models in CPN representation 
% very basic, functional, MATLAB
%
% gerwald.de 16.2.2025

% Configuration
clear
check = 0;   % 0 = sparse = scalable = fast | 1 = full = only for checks

%% Define problem

% Dimensions
n = 2;          % number of states
m = 1;          % number of inputs
p = 0;          % number of outputs
r = 6;          % rank

q = n;          % number of equations
N = 2*n+m+p;    % total number of signals

% iMTI model in CPN representation  

% S = rand(N,r);          % Structure matrix
S = [1 0 0 0 0 0; 0 1 0 0 0 0; 0 0 1 0 0 1; 0 0 0 1 0 1; 0 0 0 0 1 0];
S = sparse(S);          % Convert to sparse - un/comment for tests
% P = rand(q,r);          % Parameter matrix
P = [-1, 0, -1, 0, 1, 0; 0, -1, 0, -1, 0, 1];
% P = P.*(P<0.3);         % Set some zeros - un/comment for tests
P = sparse(P);          % Convert to sparse - un/comment for tests

% linearization point
dx = ones(n,1);         % state derivative
x = 2*ones(n,1);          % state
u = ones(m,1);          % input
y = ones(p,1);          % output
v = [dx;x;u;y];         % signal vector


tic     % start the cputime clock
%% Linearization  (Scalable = sparse & low rank)
X = S.*v-abs(S);        % Important term, better than S.*(v*ones(1,r)) 
icrit = (X==-1);        % get numerical critical indices
if any(icrit(:))
    error('specials to be implemented ?')
end
X = spfun(@(x) x+1 ,X);     % add one only to the nonzero elements
[rowi,coli,val] = find(X);  % get the indices and values of X
Y = accumarray(coli,val,[r 1],@prod)'; %' product over nonzero elements
X = spfun(@(x) 1./x, X);    % invert only nonzero elements
F = S.*Y.*X;                % compute factor matrix -> Paper
toc                         % get first cputime 

%% Full version (NOT scalable ! Only for didactics & checks)
if check 
    FC = S;                 % initialize product matrix
    X = 1-abs(S)+S.*v;      % evaluate factors 1-|S|+Sv (norm-1) 
    %X = sqrt(1-S.^2)+S.*v;  % alternative (norm-2)
    for k=1:N        % Compute products for each variable & summand
        Y = X;              % init with all factors 
        Y(k,:) = S(k,:);    % only the linear coefficient for this variable 
        FC(k,:) = prod(Y);  % compute the product over columns (lot of 1s) 
    end
end

%% Extract linear system matrices
EABC = P*F';                    %' Compute combined matrix
disp(full(EABC));
E = -EABC(:,1:n);               % dx matrix 
A = EABC(:,(n+1):2*n);          % system matrix
B = EABC(:,(2*n+1):(2*n+m));    % input matrix

%% Do analysis MATLAB based

% Local stability: dominant (geralized) eigenvalues
eigs(A,E)
% Controllability: Krylov subspace is full rank
try
    rc = rank(ctrb(A,B));
    if (rc<n)
        disp('not controllable'),
    else 
        disp('controllable');
    end
catch
    warning('not computable')
end
toc         % get first cputime 

%% Debug
if check
    disp(full([F;FC]));
end